### PR TITLE
[KIECLOUD-51] Add immutable environments

### PR DESCRIPTION
**Pull request by ruromero**
Fix https://issues.jboss.org/browse/KIECLOUD-51

Added Smartrouter as standalone object
Added omit flag to skip deploying certain resources (e.g. kieservers)
Refactored ConstructObject method
Added immutable-monitor environment
Add immutable-kieserver
Add build configuration for s2i and required data model/functionality

### CODE REVIEW

File: pkg/controller/kieapp/constants/constants.go [hunk 1]
```diff
@@ -12,3 +12,9 @@ const (
 	ImageStreamTag                    = "1.0"
 	ConfigMapPrefix                   = "kieconfigs"
 )
+
+// MonitoringEnvs Type of environments that will deploy the Monitoring console.
+// The console resources will be suffixed as -monitoring as well
+var MonitoringEnvs = map[string]struct{}{
+	"production": struct{}{},
```
Comment (bmozaffa) [inline pkg/controller/kieapp/constants/constants.go]: How about this instead?

var MonitoringEnvs = map[string]bool{
	"production": true,
}

File: pkg/controller/kieapp/defaults/defaults.go [hunk 2]
```diff
@@ -68,7 +68,7 @@ func getEnvTemplate(cr *v1.KieApp) v1.EnvTemplate {
 
 	consoleName := constants.RhpamcentrServicePrefix
 	consoleImage := constants.RhpamcentrImageName
-	if cr.Spec.KieConsoleType == v1.KieConsoleTypeMonitoring {
+	if _, isMonitoring := constants.MonitoringEnvs[cr.Spec.Environment]; isMonitoring {
```
Comment (bmozaffa) [inline pkg/controller/kieapp/defaults/defaults.go]: Would then be:

if constants.MonitoringEnvs[cr.Spec.Environment] {

File: config/common.yaml [hunk 3]
```diff
@@ -228,7 +228,126 @@ console:
...
+            imageChangeParams:
+              automatic: true
+              containerNames:
+                - "{{.ApplicationName}}-smartrouter"
+              from:
+                kind: ImageStreamTag
+                namespace: "openshift"
+                name: "rhpam{{.Version}}-smartrouter-openshift:{{.ImageTag}}"
+          - type: ConfigChange
+        replicas: 2
```
Comment (bmozaffa) [inline config/common.yaml]: Is this right / default?

File: config/envs/immutable-kieserver.yaml [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+        spec:
+          ports:
+            - port: 5432
+              targetPort: 5432
+          selector:
+            deploymentConfig: "{{.ApplicationName}}-postgresql-{{$index}}"
+    ## PostgreSQL service END
+  #{{end}}
+  ## RANGE ends
+  ## KIE Servers END
```
Comment (bmozaffa) [inline config/envs/immutable-kieserver.yaml]: Do we need to separate out the console and server for immutable env? My assumption is that we wouldn't, due to changes in the deployment model with the operator. Any initial deployment can be specified, and subsequent KIE servers can just be added to the same CR. Thoughts?

File: deploy/crs/kieapp_immutable_kieserver.yaml [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+  environment: immutable-kieserver
+  build:
+    kieServerContainerDeployment: rhpam-kieserver-library=org.openshift.quickstarts:rhpam-kieserver-library:1.4.0-SNAPSHOT
+    gitSource:
+      uri: https://github.com/jboss-container-images/rhpam-7-openshift-image.git
+      reference: master
+      contextDir: quickstarts/library-process/library
+    webhooks:
+      - type: GitHub
+        secret: secret123!
```
Comment (bmozaffa) [inline deploy/crs/kieapp_immutable_kieserver.yaml]: Would be good to always add a line break at the end of the file

File: pkg/apis/app/v1/kieapp_types.go [hunk 6]
```diff
@@ -74,22 +79,49 @@ type KieAppObject struct {
 }
 
 type Environment struct {
-	Console CustomObject   `json:"console,omitempty"`
-	Others  []CustomObject `json:"others,omitempty"`
-	Servers []CustomObject `json:"servers,omitempty"`
+	Console     CustomObject   `json:"console,omitempty"`
+	Smartrouter CustomObject   `json:"smartrouter,omitempty"`
+	Servers     []CustomObject `json:"servers,omitempty"`
```
Comment (bmozaffa) [inline pkg/apis/app/v1/kieapp_types.go]: I haven't fully thought this through yet, but what does it mean for omit to be on an individual CustomObject, whereas there is an array of them?

File: pkg/controller/kieapp/defaults/defaults.go [hunk 7]
```diff
@@ -61,24 +61,27 @@ func getEnvTemplate(cr *v1.KieApp) v1.EnvTemplate {
...
-		cr.Spec.Objects.Server.Env = []corev1.EnvVar{{Name: "empty"}}
-	}
 	if cr.Spec.RhpamRegistry == (v1.KieAppRegistry{}) {
 		cr.Spec.RhpamRegistry.Registry = logs.GetEnv("REGISTRY", constants.RhpamRegistry) // default to red hat registry
 		cr.Spec.RhpamRegistry.Insecure = logs.GetBoolEnv("INSECURE")
 	}
 
+	consoleName := constants.RhpamcentrServicePrefix
+	consoleImage := constants.RhpamcentrImageName
+	if _, isMonitoring := constants.MonitoringEnvs[cr.Spec.Environment]; isMonitoring {
```
[the remark this answers is not in the document]
Comment (ruromero) [reply pkg/controller/kieapp/defaults/defaults.go]: that looks simpler, yes

File: config/common.yaml (hunk 3 above)
Comment (ruromero) [reply config/common.yaml]: Yes, it is the common value for prod and immutable-monitor

File: pkg/apis/app/v1/kieapp_types.go (hunk 6 above)
Comment (ruromero) [reply pkg/apis/app/v1/kieapp_types.go]: That'd imply to add a new object to have omit + customObject and another one for omit + []customObject. I didn't want to complicate much the current data structure.
```
env.Servers.Omit // bool
env.Servers.Objects[i] // CustomObject
env.Console.Omit // bool
env.Console.Object // CustomObject
```

**Comment by ruromero** [comment]
@bmozaffa I am getting the following error while trying to create the BuildConfig:
```
{"level":"error","ts":1546520266.2494493,"logger":"kieapp.controller","msg":"Failed to get object. no kind is registered for the type v1.BuildConfig in scheme \"k8s.io/client-go/kubernetes/scheme/register.go:60\"","kind":"BuildConfig","name":"immutablekie-kieserver-0","namespace":"myproject"}
{"level":"error","ts":1546520266.2494972,"logger":"kubebuilder.controller","msg":"Reconciler error","controller":"kieapp-controller","request":"myproject/immutablekie","error":"no kind is registered for the type v1.BuildConfig in scheme \"k8s.io/client-go/kubernetes/scheme/register.go:60\"","stacktrace":"github.com/kiegroup/kie-cloud-operator/vendor/github.com/go-logr/zapr.(*zapLogger).Error\n\t/home/rromerom/workspaces/go/src/github.com/kiegroup/kie-cloud-operator/vendor/github.com/go-logr/zapr/zapr.go:128\ngithub.com/kiegroup/kie-cloud-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/home/rromerom/workspaces/go/src/github.com/kiegroup/kie-cloud-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:215\ngithub.com/kiegroup/kie-cloud-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1\n\t/home/rromerom/workspaces/go/src/github.com/kiegroup/kie-cloud-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:158\ngithub.com/kiegroup/kie-cloud-operator/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1\n\t/home/rromerom/workspaces/go/src/github.com/kiegroup/kie-cloud-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133\ngithub.com/kiegroup/kie-cloud-operator/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil\n\t/home/rromerom/workspaces/go/src/github.com/kiegroup/kie-cloud-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134\ngithub.com/kiegroup/kie-cloud-operator/vendor/k8s.io/apimachinery/pkg/util/wait.Until\n\t/home/rromerom/workspaces/go/src/github.com/kiegroup/kie-cloud-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88"}

```

Do you know if it is related to the fact that you cannot create BuildConfigs directly?
I am creating it in the same way as the other resources see: https://github.com/kiegroup/kie-cloud-operator/pull/36/commits/2d901fc2addf6be31a36666646362a8fe7155c06#diff-9ecb910ca7304714ab23ba15d8cf55a5R524

**Pull request merged**